### PR TITLE
Re-work of Terraform scripts to use latest versions and provide three working examples

### PR DESCRIPTION
**Pull request by andyrepton**
I've split the examples into three sections:

1. Managed OIDC provider with no modules in between. This is useful for people getting started with ROSA and Terraform, and is easy enough to copy and paste from multiple files. This came out of customers not wanting to have to copy and paste multiple modules from our github (for example those who could not git-clone from github to their TFE setup)
2. Managed OIDC provider with modules and a single main.tf file. This is a cleaner setup and is meant as a starting place for most people
3. Unmanaged OIDC provider with modules and a single main.tf file. The same but for those who wish to use an unmanaged OIDC provider

Fixes:
- Removed any unneeded and unclear options, for example where a managed OIDC provider is used, removed any extra sections not required to make it cleaner to read
- Fixed two race conditions where account_roles were not available yet before attempting to create either the ROSA cluster or the OIDC provider in the case of the unmanaged version

Improvements:
- Moved account_role creation, OIDC_provider creation and Operator roles creation into their own module calls, to make it clear how these can happen separately. It is an exercise for the reader to combine them if they wish to, but I feel this affects readability
- Removed the `var.token` option and documented to use the environment variable RHCS_TOKEN as a best practise to avoid leaking the ROSA cluster token
- Formatted all files to be consistent

Plus more! I have tested a full terraform create/destroy in each folder as of 1st November 2023.

### CODE REVIEW

File: rosa_sts_managed_oidc/main.tf [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+# is created. Hopefully this is fixed in the future
+resource "time_sleep" "wait_10_seconds" {
+  depends_on = [module.account_role]
+
+  create_duration = "10s"
+}
+
+# Create managed OIDC config
+module "oidc_provider" {
+  source               = "../modules/managed_oidc_provider"
```
Comment (bardielle) [inline rosa_sts_managed_oidc/main.tf:53]: I think it's better to add here the depends_on :
`depends_on = [time_sleep.wait_10_seconds]` 
So the oidc provider will start the creation after 10s delay

Comment (andyrepton) [reply rosa_sts_managed_oidc/main.tf:53]: This isn't needed for the managed oidc provider as it doesn't depend on the account roles as far as I can tell. If you feel there's a risk of a race condition here we can also add a depends_on similar to the example in the unmanaged_oidc_provider example

Comment (bardielle) [reply rosa_sts_managed_oidc/main.tf:53]: I think it's a good practice... 
Did you test it and it works?

File: rosa_sts_managed_oidc/main.tf [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  autoscaling_enabled    = var.autoscaling_enabled
+  min_replicas           = var.min_replicas
+  max_replicas           = var.max_replicas
+  oidc_config_id         = module.oidc_provider.id
+  additional_tags        = var.additional_tags
+  vpc_cidr_block         = var.vpc_cidr_block
+
+  #private link cluster values
+  enable_private_link = var.enable_private_link
+  aws_subnet_ids   = var.create_vpc ? module.vpc.private_subnets : var.aws_subnet_ids
```
Comment (bardielle) [inline rosa_sts_managed_oidc/main.tf:93]: AFAK you should set also public subnets.... not only private...

File: rosa_sts_managed_oidc/main.tf (hunk 1 above)
Comment (andyrepton) [reply rosa_sts_managed_oidc/main.tf:53]: Added. Yes, I've done full `terraform init; terraform plan; terraform apply; terraform destroy` with all of these

File: rosa_sts_managed_oidc/main.tf (hunk 2 above)
Comment (andyrepton) [reply rosa_sts_managed_oidc/main.tf:93]: Currently this isn't working with public clusters. It's something that needs to be edited. I've been focusing on Black Belt customer use cases which are almost 100% of the time private clusters. Once I have some free time I'll come back to this and add tests for public, or someone else is more than welcome to add this in with the appropriate check if the cluster is not private :-)

File: modules/rosa_cluster/variables.tf [hunk 3]
```diff
@@ -23,23 +18,23 @@ variable "availability_zones" {
...
+  default = "kumudu"
 }
 
-variable operator_role_prefix {
-    type = string
-    default = "mobbkh"
+variable "operator_role_prefix" {
+  type    = string
+  default = "mobbkh"
 }
```
Comment (michaelryanmcneill) [inline modules/rosa_cluster/variables.tf]: Do we want to consider changing these to non-specific vars?

File: rosa_sts_managed_oidc_no_modules/terraform.tfvars [hunk 4]
```diff
@@ -0,0 +1,38 @@
+#AWS Info
+aws_region="eu-west-1"
+availability_zones = ["eu-west-1a","eu-west-1b","eu-west-1c"]
+
+# Create a VPC?
+create_vpc=true
+vpc_name="poc-andyr1"
```
Comment (michaelryanmcneill) [inline rosa_sts_managed_oidc_no_modules/terraform.tfvars:7]: Same question here, do we want to specify non-user specific vars?

File: rosa_sts_managed_oidc_no_modules/variables.tf [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  type        = bool
+  description = "Red Hat managed or unmanaged (Customer hosted) OIDC Configuration"
+  default     = true
+}
+
+# ROSA Cluster info
+variable "cluster_name" {
+  type        = string
+  description = "The name of the ROSA cluster to create"
+  default     = "poc-andyr"
```
Comment (michaelryanmcneill) [inline rosa_sts_managed_oidc_no_modules/variables.tf:58]: Same question here, do we want to specify non-user specific vars?

File: rosa_sts_unmanaged_oidc/variables.tf [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    sts_instance_controlplane_permission_policy = string
+  })
+  default = null
+}
+
+# Used ?
+variable "rh_oidc_provider_thumbprint" {
+  description = "Thumbprint for https://rh-oidc.s3.us-east-1.amazonaws.com"
+  type        = string
+  default     = "917e732d330f9a12404f73d8bea36948b929dffc"
```
Comment (michaelryanmcneill) [inline rosa_sts_unmanaged_oidc/variables.tf]: Do we want to specify a default here?

File: rosa_sts_unmanaged_oidc/variables.tf [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+variable "aad_app_password_name" {
+  type        = string
+  default     = "kh-ROSA-Secret"
+  description = "AAD Secret Name"
+}
+
+variable "aad_app_redirect_uri" {
+  type        = string
+  default     = "https://oauth-openshift.apps.iv2wbc7f.westus2.aroapp.io/oauth2callback/AAD"
```
Comment (michaelryanmcneill) [inline rosa_sts_unmanaged_oidc/variables.tf:225]: We should remove this default.

**Review by michaelryanmcneill — APPROVED** [review]
LGTM, only thing I would consider is possibly changing the `variables.tf` files to be a `variables.tf.sample` or something like that. Seems there is some user-specific content that shouldn't be in examples.

File: rosa_sts_unmanaged_oidc/variables.tf (hunk 6 above)
Comment (andyrepton) [reply rosa_sts_unmanaged_oidc/variables.tf]: Removing!

File: rosa_sts_unmanaged_oidc/variables.tf (hunk 7 above)
Comment (andyrepton) [reply rosa_sts_unmanaged_oidc/variables.tf:225]: The whole thing isn't used right now, but indeed

File: rosa_sts_managed_oidc_no_modules/variables.tf (hunk 5 above)
Comment (andyrepton) [reply rosa_sts_managed_oidc_no_modules/variables.tf:58]: No, I'll make it a prompt

File: rosa_sts_managed_oidc_no_modules/terraform.tfvars (hunk 4 above)
Comment (andyrepton) [reply rosa_sts_managed_oidc_no_modules/terraform.tfvars:7]: Nope, I'll fix this

File: modules/rosa_cluster/variables.tf (hunk 3 above)
Comment (andyrepton) [reply modules/rosa_cluster/variables.tf]: Yes indeed

**Pull request merged**